Traverse sub-blocks in JIT passes

apaszke opened the pull request:
Haven't touched the ONNX pass. I'm not up to date with the plans/node definitions for exporting control flow, so I'll leave that to others.

apaszke committed:
Traverse sub-blocks in JIT passes

zdevito commented on torch/csrc/jit/passes/graph_fuser.cpp:
```diff
@@ -494,19 +493,24 @@ struct GraphFuser {
     }
+    for (Node * node : block->nodes()) {
+      for (Block * sub_block : node->blocks()) {
+        GraphFuser(sub_block).run();
```
What stops nodes from fusing across blocks? Fusing across blocks is not always safe since it may try to move a producer into a consumer that is only executed conditionally (or executed multiple times, which is a perf bug). 
I believe we need to add a check into all `isFusable` functions (and other places where we move instructions around) that the two operations are in the same blocks.

apaszke replied:
Hmmm yeah, I think I checked that it was safe in BatchMM, and then concluded that I'm happy with "every block-level function has independent state, and works on nodes from a single block, so it should be ok", but yeah, it's obviously false when we're trying to traverse the graph via inputs.

zdevito commented on torch/csrc/jit/passes/peephole.cpp:
```diff
@@ -40,7 +40,15 @@ void PeepholeOptimize(std::shared_ptr<Graph>& graph) {
     }
+
+    for (Block * sub_block : n->blocks()) {
+      PeepholeOptimize(sub_block);
```
Similar comment here to fuser: if peephole optimization previously assumed everything was in the same block, those checks need to be added now.

zdevito reviewed (commented):
Making these changes is a good idea, but I am concerned about correctness in the Fuser and Peephole optimizer. Those passes were written assuming no control flow, so simply scanning all the blocks could cause fusions to happen that are no longer safe. This needs to have some test cases to make sure a unsafe fusion across two blocks does not occur.

apaszke committed:
Add an extra check to prevent cross-block fusion

apaszke committed:
Add comment

apaszke commented:
I've changed the fuser to make it treat nodes in different blocks as not fusable. Regarding peephole, I don't think any of the optimizations we have at the moment require any changes, because the only:
1. replace values produced in a current block with values form previous blocks
2. do 1. and remove nodes from previous blocks if and only if they are no longer used anywhere else.

I've added a comment mentioning that if someone was to make a simplification that would replace a subgraph of some kind, then they will need to take extra care to ensure that all nodes are in the same block.
